### DISCUSSION TITLE
How to use a uarray.array object in C?

### DISCUSSION THREAD

**Discussion by TsXor** [micropython/micropython#9823] [Using MicroPython]
There seems to be no present api.

**Comment by dlech** [comment]
Generally, for any object that implements the buffer protocol (bytes, bytearray, array, etc.), you use `mp_get_buffer_raise()` to get the underlying C buffer from the object.

**Reply by TsXor** [reply]
How can I know the type of an array? Property typecode cannot be accessed.

**Reply by jimmo** [reply]
The buffer knows its type code

See for example:
https://github.com/micropython/micropython/blob/master/extmod/modframebuf.c#L582

**Reply by TsXor** [reply]
Then how can I cast the buffer to the target type? As far as I know, C does not have a flexible type system like Python.

**Reply by dlech** [reply]
You could use a switch statement if you need to handle multiple possible types. Or you could check for a specific type and raise an exception for other types.

What exactly are you trying to accomplish?

**Reply by TsXor** [reply]
> You could use a switch statement if you need to handle multiple possible types. Or you could check for a specific type and raise an exception for other types.
> 
> What exactly are you trying to accomplish?

Well I did it, but now I have another question: how to "transfer the ownership of" a buffer so that it is free from GC and can be freed at C side?

**Reply by jimmo** [reply]
> I have another question: how to "transfer the ownership of" a buffer so that it is free from GC and can be freed at C side?

There's no real concept of ownership, only that if something points to the allocation then the GC won't free it.

If it was allocated on the MicroPython heap then it can't be transferred out of the heap. It will be free'd automatically when nothing references it.

If you're sure that nothing else is referencing it, then you may call `m_free` (or `m_del`) to force the GC to free it immediately. You should ensure that any pointers are set to NULL (so that they don't accidentally point to a future allocation at that address).

**Reply by TsXor** [reply]
> > I have another question: how to "transfer the ownership of" a buffer so that it is free from GC and can be freed at C side?
> 
> There's no real concept of ownership, only that if something points to the allocation then the GC won't free it.
> 
> If it was allocated on the MicroPython heap then it can't be transferred out of the heap. It will be free'd automatically when nothing references it.
> 
> If you're sure that nothing else is referencing it, then you may call `m_free` (or `m_del`) to force the GC to free it immediately. You should ensure that any pointers are set to NULL (so that they don't accidentally point to a future allocation at that address).

Now I am really curious WHAT on earth is "referencing"?
Let me guess: all variables and attributes in uPy are qstr-s, and "referenced" just means that a pointer (yes, mp_obj_t) to uPy instance struct is somehow associated with a qstr?

**Reply by jimmo** [reply]
> Now I am really curious WHAT on earth is "referencing"?
> Let me guess: all variables and attributes in uPy are qstr-s, and "referenced" just means that a pointer (yes, mp_obj_t) to uPy instance struct is somehow associated with a qstr?

I'm not quite sure what you're referring here with the qstrs. qstrs are just interned strings which provides a convenient way to encode string references in the bytecode as well as de-duplicating strings. Note that interned strings are never garbage collected -- the QSTR pool is append-only.

Something is "referenced" in MicroPython if there is live pointer to that object. A live pointer is a pointer that can be found in memory starting at either any one of the root pointers, the stack, or any CPU registers, that points to a GC object. Once a live object is found, all word-aligned pointers in the memory of that object are then tried...and recursively so. (It's very simple! Yes it does mean sometimes you end up marking an object live that isn't, but you never miss anything).

**Reply by TsXor** [reply]
> I'm not quite sure what you're referring here with the qstrs. qstrs are just interned strings which provides a convenient way to encode string references in the bytecode as well as de-duplicating strings. Note that interned strings are never garbage collected -- the QSTR pool is append-only.
> 
> Something is "referenced" in MicroPython if there is live pointer to that object. A live pointer is a pointer that can be found in memory starting at either any one of the root pointers, the stack, or any CPU registers, that points to a GC object. Once a live object is found, all word-aligned pointers in the memory of that object are then tried...and recursively so. (It's very simple! Yes it does mean sometimes you end up marking an object live that isn't, but you never miss anything).

Sorry about that: I was describing refcount, which is in CPython.

**Reply by TsXor** [reply]
> Something is "referenced" in MicroPython if there is live pointer to that object. A live pointer is a pointer that can be found in memory starting at either any one of the root pointers, the stack, or any CPU registers, that points to a GC object. Once a live object is found, all word-aligned pointers in the memory of that object are then tried...and recursively so. (It's very simple! Yes it does mean sometimes you end up marking an object live that isn't, but you never miss anything).

```c
mp_obj_t ptr_pool[0];
mp_obj_new_list(0, ptr_pool);
mp_store_global(MP_QSTR__ptr_pool, MP_OBJ_FROM_PTR(&ptr_pool));
```
and I put some pointers into ptr_pool with `mp_obj_list_append`, can this action keeps the data that these pointers point to from being recycled?

**Reply by dlech** [reply]
In the code you shared, `ptr_pool` is allocated on the stack instead of from the heap so will no longer be valid after the function where it was allocated exits. Also, it has zero size, so it isn't really allocating anything.

I think what you meant instead is:
```c
mp_obj_t obj_pool = mp_obj_new_list(0, NULL);
mp_store_global(MP_QSTR__obj_pool, obj_pool);
```

You can also use `MP_REGISTER_ROOT_POINTER` instead if you have a fixed number of objects to keep track of and don't want to expose the object in the Python globals.

**Reply by TsXor** [reply]
> Something is "referenced" in MicroPython if there is live pointer to that object. A live pointer is a pointer that can be found in memory starting at either any one of the root pointers, the stack, or any CPU registers, that points to a GC object. Once a live object is found, all word-aligned pointers in the memory of that object are then tried...and recursively so. (It's very simple! Yes it does mean sometimes you end up marking an object live that isn't, but you never miss anything).

I thought about it again, can "reference" be simpler, like this:
```c
// globally
# define POOL_INIT_SIZE 64
mp_obj_t ptr_pool = (mp_obj_t)m_malloc(POOL_INIT_SIZE * sizeof(mp_obj_t));

// in init function
mp_store_global(MP_QSTR__ptr_pool, ptr_pool);

// in some functions
ptr_pool[x] = ptr;
```

When a part of memory can be found starting from a root pointer, it is not target of GC?
```
void *root_ptr ----> void *ptr1 ----> void *ptr2 ----> ... ----> void *ptrn ---> some memory
```

**Reply by TsXor** [reply]
I'm not using `MP_REGISTER_ROOT_POINTER` because I'm using the lvgl version, and the result of a lookup showed that it lacked `py/make_root_pointers.py`, which parses this statement.

**Reply by jimmo** [reply]
> I thought about it again, can "reference" be simpler, like this:

Yes. But this is exactly what `m_tracked_calloc` does, so I'd recommend using that instead.

> When a part of memory can be found starting from a root pointer, it is not target of GC?

Yes exactly.

> I'm not using MP_REGISTER_ROOT_POINTER because

In that case you should use `MICROPY_PORT_ROOT_POINTERS` in mpconfigport.h (this was replaced by MP_REGISTER_ROOT_POINTER) or `m_tracked_calloc`.
